server/src/main/java/org/elasticsearch/cluster/metadata/FieldInferenceMetadata.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.Diffable;
+import org.elasticsearch.cluster.DiffableUtils;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.mapper.MappingLookup;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Contains field inference information. This is necessary to add to cluster state as inference can be calculated in the coordinator
+ * node, which not necessarily has mapping information.
+ */
+public class FieldInferenceMetadata implements Diffable<FieldInferenceMetadata>, ToXContentFragment {
+
+    // Keys: field names. Values: Inference ID associated to the field name for calculating inference
+    private final ImmutableOpenMap<String, String> inferenceIdForField;
+
+    // Keys: field names. Values: Field names that provide source for this field (either as copy_to or multifield sources)
+    private final ImmutableOpenMap<String, Set<String>> sourceFields;
```

Comment:
We may want to change this to `ImmutableOpenMap<String, List<String>>` so that the source field iteration order is consistent

Reply:
I don't think order is of consequence. `FieldTypeLookup.sourcePaths()`, which is used for similar purposes, also returns a `Set`.

Reply:
I bring it up because @benwtrent brought it up during multi-field/copy_to support discovery. Depending on how we handle inference generation for multi-valued fields in the bulk action, order could matter. 

For example, if we handle it by concatenating all the values, order could affect the chunks that are produced and therefore the inference results.